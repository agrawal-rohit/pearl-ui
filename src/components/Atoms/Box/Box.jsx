import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
