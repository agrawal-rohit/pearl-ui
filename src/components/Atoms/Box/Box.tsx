import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
