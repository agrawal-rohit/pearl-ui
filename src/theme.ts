import { createTheme } from "@shopify/restyle";

const palette = {
  purpleLight: "#8C6FF7",
  purplePrimary: "#5A31F4",
  purpleDark: "#3F22AB",

  greenLight: "#56DCBA",
  greenPrimary: "#0ECD9D",
  greenDark: "#0A906E",

  black: "#0B0B0B",
  white: "#F0F2F3",
};

const theme = createTheme({
  colors: {
    primary: palette.purplePrimary,
    ...palette,
  },
  spacing: {
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
  },
  breakpoints: {
    phone: 0,
    tablet: 768,
  },
});

export type Theme = typeof theme;
export default theme;
