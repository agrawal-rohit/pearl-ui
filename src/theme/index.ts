// Theme functions
export { createTheme, extendTheme } from "./src/themeFunctions";

// Style functions
export {
  color,
  opacity,
  backgroundColor,
  spacing,
  typography,
  layout,
  position,
  border,
  shadow,
  textShadow,
} from "./src/styleFunctions";

// Theme Context
export { ThemeProvider } from "./src/themeContext";
export { baseTheme } from "./src/basetheme";

// Utils
export { generatePalette } from "./utils/utils";

// Types
export { BasePearlTheme as PearlTheme } from "./src/types";
