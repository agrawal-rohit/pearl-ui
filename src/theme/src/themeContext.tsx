import AppLoading from "expo-app-loading";
import {
  useFonts,
  Poppins_300Light,
  Poppins_400Regular,
  Poppins_500Medium,
  Poppins_600SemiBold,
  Poppins_700Bold,
  Poppins_800ExtraBold,
} from "@expo-google-fonts/poppins";
import React, { useEffect, useState, createContext } from "react";
import { useColorScheme } from "react-native";
import { baseTheme } from "./basetheme";
import { BasePearlTheme } from "./types";

type ThemeType = "light" | "dark" | "system";

export interface IThemeContext {
  /** Theme configuration object for the active color mode */
  theme: BasePearlTheme;
  /** Active color mode */
  colorMode: ThemeType;
  /** Function to toggle the active color mode */
  toggleColorMode(): void;
}

interface ThemeProviderProps {
  /** Initial color mode for the theme (light, dark, system) */
  initialColorMode?: ThemeType;
  /** The theme configuration object */
  theme?: BasePearlTheme;
  /** A mapping object of the fonts used by the theme */
  fonts?: object;
  /**React children */
  children: React.ReactNode;
}

const baseFonts = {
  "Poppins-Light": Poppins_300Light,
  "Poppins-Regular": Poppins_400Regular,
  "Poppins-Medium": Poppins_500Medium,
  "Poppins-SemiBold": Poppins_600SemiBold,
  "Poppins-Bold": Poppins_700Bold,
  "Poppins-ExtraBold": Poppins_800ExtraBold,
};

export const themeContext = createContext({} as IThemeContext);

/**
 * The main provider component which controls the theme and color mode of the components in the entire application.
 */
export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  initialColorMode = "light",
  theme = baseTheme,
  fonts = baseFonts,
  children,
}) => {
  const [colorMode, setColorMode] = useState<ThemeType>(initialColorMode);
  const systemThemeStyle = useColorScheme() as ThemeType;
  const [haveFontsLoaded] = useFonts(fonts);

  const changeThemeTo = (colorMode: ThemeType) => {
    if (colorMode === "dark") {
      setColorMode("dark");
    } else {
      setColorMode("light");
    }
  };

  const toggleColorMode = () => {
    if (colorMode === "light") {
      changeThemeTo("dark");
    } else {
      changeThemeTo("light");
    }
  };

  useEffect(() => {
    // Set initial Theme
    if (colorMode === "light") {
      changeThemeTo("light");
    } else if (colorMode === "dark") {
      changeThemeTo("dark");
    } else if (colorMode === "system") {
      changeThemeTo(systemThemeStyle);
    }
  }, [colorMode, initialColorMode, theme]);

  if (haveFontsLoaded) {
    return (
      <themeContext.Provider
        value={{ theme, colorMode, toggleColorMode } as IThemeContext}
      >
        {children}
      </themeContext.Provider>
    );
  }

  return <AppLoading />;
};
