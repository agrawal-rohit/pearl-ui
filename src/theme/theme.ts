import { createTheme } from "@shopify/restyle";
import namedColors from "./namedColors.json";

export interface IColor {
  [key: string]: string;
}

export const palette = {
  primary100: "#E1E6FF",
  primary200: "#C3CCFF",
  primary300: "#A5B1FF",
  primary400: "#8F9DFF",
  primary500: "#6A7BFF",
  primary600: "#4D5BDB",
  primary700: "#3541B7",
  primary800: "#212A93",
  primary900: "#141B7A",

  secondary100: "#FFE1E7",
  secondary200: "#FFC3D6",
  secondary300: "#FFA5CB",
  secondary400: "#FF8FC9",
  secondary500: "#FF6AC6",
  secondary600: "#DB4DB3",
  secondary700: "#B7359F",
  secondary800: "#93218A",
  secondary900: "#78147A",

  neutral100: "#FFFFFF",
  neutral200: "#F7F9FC",
  neutral300: "#EDF1F7",
  neutral400: "#E4E9F2",
  neutral500: "#C5CEE0",
  neutral600: "#8F9BB3",
  neutral700: "#2E3A59",
  neutral800: "#222B45",
  neutral900: "#1A2138",

  success100: "#F5FCCC",
  success200: "#EAFA9A",
  success300: "#D6F266",
  success400: "#C0E640",
  success500: "#A2D608",
  success600: "#86B805",
  success700: "#6C9A04",
  success800: "#537C02",
  success900: "#426601",

  info100: "#D8FEFB",
  info200: "#B2FEFC",
  info300: "#8CF6FC",
  info400: "#6FE8FA",
  info500: "#40D2F7",
  info600: "#2EA6D4",
  info700: "#207EB1",
  info800: "#145B8F",
  info900: "#0C4176",

  warning100: "#FFF7D6",
  warning200: "#FFEDAE",
  warning300: "#FFE085",
  warning400: "#FFD367",
  warning500: "#FFBF35",
  warning600: "#DB9C26",
  warning700: "#B77B1A",
  warning800: "#935D10",
  warning900: "#7A480A",

  danger100: "#FFE7D9",
  danger200: "#FFCAB3",
  danger300: "#FFA68D",
  danger400: "#FF8371",
  danger500: "#FF4B42",
  danger600: "#DB3036",
  danger700: "#B72133",
  danger800: "#93152E",
  danger900: "#7A0C2C",

  ...namedColors,
};

export const baseLightTheme = createTheme({
  colors: {
    ...palette,

    mainBackground: palette.neutral100,
    textColor: palette.neutral800,
    buttonTextColor: palette.neutral100,
  },
  spacing: {
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
  },
  breakpoints: {
    phone: 0,
    longPhone: {
      width: 0,
      height: 812,
    },
    tablet: 768,
    largeTablet: 1024,
  },
  textVariants: {
    h1: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 42,
      color: "textColor",
    },
    h2: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 36,
      color: "textColor",
    },
    h3: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 32,
      color: "textColor",
    },
    h4: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 30,
      color: "textColor",
    },
    h5: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 26,
      color: "textColor",
    },
    h6: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 22,
      color: "textColor",
    },
    s1: {
      fontFamily: "Poppins-SemiBold",
      fontWeight: "600",
      fontSize: 22,
      color: "textColor",
    },
    s2: {
      fontFamily: "Poppins-SemiBold",
      fontWeight: "600",
      fontSize: 19,
      color: "textColor",
    },
    p1: {
      fontFamily: "Poppins-Regular",
      fontWeight: "400",
      fontSize: 19,
      color: "textColor",
    },
    p2: {
      fontFamily: "Poppins-Regular",
      fontWeight: "400",
      fontSize: 17,
      color: "textColor",
    },
    c1: {
      fontFamily: "Poppins-Regular",
      fontWeight: "400",
      fontSize: 17,
      color: "textColor",
    },
    c2: {
      fontFamily: "Poppins-Medium",
      fontWeight: "500",
      fontSize: 17,
      color: "textColor",
    },
    label: {
      fontFamily: "Poppins-Bold",
      fontWeight: "bold",
      fontSize: 17,
      color: "textColor",
    },
    button: {
      fontFamily: "Poppins-Medium",
      fontWeight: "500",
      fontSize: 17,
      color: "buttonTextColor",
    },
    notifications: {
      fontFamily: "Poppins-Medium",
      fontWeight: "500",
      fontSize: 17,
      color: "buttonTextColor",
    },
  },
  activityIndicatorSize: {
    s: {
      size: 10,
    },
    m: {
      size: 30,
    },
    l: {
      size: 50,
    },
    xl: {
      size: 70,
    },
  },
});

export type Theme = typeof baseLightTheme;

export const baseDarkTheme: Theme = {
  ...baseLightTheme,
  colors: {
    ...baseLightTheme.colors,
    mainBackground: palette.neutral700,
    textColor: palette.neutral100,
  },
};
