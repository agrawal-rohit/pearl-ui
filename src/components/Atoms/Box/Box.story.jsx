import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "./Box";

storiesOf("Box", module)
  .addDecorator((getStory) => getStory())
  .add("Basic box", () => (
    <Box
      padding="s"
      height={300}
      width={100}
      backgroundColor="primary100"
    ></Box>
  ));
