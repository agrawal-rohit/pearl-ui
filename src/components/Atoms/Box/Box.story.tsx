import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Box", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Basic box", () => (
    <Box padding="s" width={200} backgroundColor="primary100"></Box>
  ));
