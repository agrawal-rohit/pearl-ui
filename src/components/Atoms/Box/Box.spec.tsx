import React from "react";
import Box from "./Box";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "../../../theme/themeContext";

jest.useFakeTimers();

describe("Atoms/Box", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <Box>asdasd</Box>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
