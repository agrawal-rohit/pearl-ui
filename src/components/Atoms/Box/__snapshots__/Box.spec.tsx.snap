// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Box passes the snapshot test 1`] = `
<View
  style={
    Object {
      "display": "flex",
    }
  }
>
  asdasd
</View>
`;
